test/src/main/java/ai/timefold/solver/test/impl/score/stream/DefaultShadowVariableAwareMultiConstraintAssertion.java

```diff
@@ -0,0 +1,73 @@
+package ai.timefold.solver.test.impl.score.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+import ai.timefold.solver.core.impl.score.constraint.ConstraintMatchPolicy;
+import ai.timefold.solver.core.impl.score.director.stream.BavetConstraintStreamScoreDirector;
+import ai.timefold.solver.core.impl.score.stream.common.AbstractConstraintStreamScoreDirectorFactory;
+import ai.timefold.solver.test.api.score.stream.MultiConstraintAssertion;
+import ai.timefold.solver.test.api.score.stream.ShadowVariableAwareMultiConstraintAssertion;
+
+public final class DefaultShadowVariableAwareMultiConstraintAssertion<Solution_, Score_ extends Score<Score_>>
+        extends AbstractMultiConstraintAssertion<Score_> implements ShadowVariableAwareMultiConstraintAssertion {
+
+    private final AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_, ?> scoreDirectorFactory;
+    private final Solution_ solution;
+    private boolean initialized = false;
+
+    DefaultShadowVariableAwareMultiConstraintAssertion(ConstraintProvider constraintProvider,
+            AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_, ?> scoreDirectorFactory,
+            Solution_ solution) {
+        super(constraintProvider);
+        this.scoreDirectorFactory = requireNonNull(scoreDirectorFactory);
+        this.solution = Objects.requireNonNull(solution);
+    }
+
+    @Override
+    void ensureInitialized() {
+        if (!initialized) {
```

Comment:
In case of long blocks of code, I find that this reads better:

```suggestion
        if (initialized) {
            return;
        }
        // Do the rest not nested. No longer need to think about what the condition was.
        ...
```